File: python_rewrite/steering_functions/configuration.py
"""Port of hc_cc_core/configuration.hpp and configuration.cpp."""

import math

from steering_functions.utilities import twopify, point_distance, get_epsilon


class Configuration:
    """Kinematic configuration (position, orientation, curvature)."""

    def __init__(self, x=0.0, y=0.0, theta=0.0, kappa=0.0):
        self.x = x
        self.y = y
        self.theta = twopify(theta)
        self.kappa = kappa

    def __repr__(self):
        return (
            f"Configuration({self.x}, {self.y}, {self.theta}, {self.kappa})"
        )


def configuration_distance(q1, q2):
    """Cartesian distance between two configurations."""
    return point_distance(q1.x, q1.y, q2.x, q2.y)


def configuration_aligned(q1, q2):
    """Check whether two configurations are aligned."""
    if abs(q2.theta - q1.theta) > get_epsilon():
        return False
    angle = twopify(math.atan2(q2.y - q1.y, q2.x - q1.x))
    return abs(angle - q1.theta) <= get_epsilon()


def configuration_equal(q1, q2):
    """Check whether two configurations are equal."""
    if abs(q2.theta - q1.theta) > get_epsilon():
        return False
    if configuration_distance(q1, q2) > get_epsilon():
        return False
    return True

File: rust/src/configuration.rs
use crate::utilities::{twopify, point_distance, get_epsilon};

/// Kinematic configuration: position, orientation, curvature.
#[derive(Clone, Copy, Debug, Default)]
pub struct Configuration {
    pub x: f64,
    pub y: f64,
    pub theta: f64,
    pub kappa: f64,
}

impl Configuration {
    pub fn new(x: f64, y: f64, theta: f64, kappa: f64) -> Self {
        Self { x, y, theta: twopify(theta), kappa }
    }
}

pub fn configuration_distance(q1: &Configuration, q2: &Configuration) -> f64 {
    point_distance(q1.x, q1.y, q2.x, q2.y)
}

pub fn configuration_aligned(q1: &Configuration, q2: &Configuration) -> bool {
    if (q2.theta - q1.theta).abs() > get_epsilon() {
        return false;
    }
    let angle = twopify((q2.y - q1.y).atan2(q2.x - q1.x));
    (angle - q1.theta).abs() <= get_epsilon()
}

pub fn configuration_equal(q1: &Configuration, q2: &Configuration) -> bool {
    if (q2.theta - q1.theta).abs() > get_epsilon() {
        return false;
    }
    configuration_distance(q1, q2) <= get_epsilon()
}
